[clang-doc] Handle static members and functions

ilovepi opened the pull request:
clang-doc didn't visit VarDecl, and hence never collected info
from class statics members and functions.

Fixes #59813.

ilovepi committed:
[clang-doc] Handle static members and functions

clang-doc didn't visit VarDecl, and hence never collected info
from class statics members and functions.

Fixes #59813.

ilovepi commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/135457?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#135457** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/135457?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/135457?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#135456** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/135456?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

petrhosek commented on clang-tools-extra/clang-doc/Serialize.cpp:
```diff
@@ -729,6 +729,62 @@ emitInfo(const RecordDecl *D, const FullComment *FC, int LineNumber,
+                 dyn_cast<ClassTemplatePartialSpecializationDecl *>(SpecOf))
+      Specialization.SpecializationOf = getUSRForDecl(CTPSD);
+
+    // Parameters to the specilization. For partial specializations, get the
```
```suggestion
    // Parameters to the specialization. For partial specializations, get the
```

ilovepi committed:
Fix typo

ilovepi committed:
Fix file separator in test for Windows

ilovepi committed:
Allow for escaped forward slash in Windows paths

ilovepi committed:
Fix path check in MD-CALC

ilovepi commented on clang-tools-extra/clang-doc/Serialize.cpp:
```diff
@@ -687,6 +687,62 @@ emitInfo(const RecordDecl *D, const FullComment *FC, int LineNumber,
+std::pair<std::unique_ptr<Info>, std::unique_ptr<Info>>
+emitInfo(const VarDecl *D, const FullComment *FC, int LineNumber,
+         llvm::StringRef File, bool IsFileInRootDir, bool PublicOnly) {
+  auto I = std::make_unique<RecordInfo>();
```
This probably should be a different type, instead of a Record. I'll experiment some and see if we have an existing Info class that will work a bit better.

ilovepi committed:
Mark static functions and members in normal processing

Visiting VarDecls looses context. Instead, track additional information
about statics within the Info* classes, and update serialization to
handle it. We can visit VarDecl's as a normal part of visiting Records,
and keep the necessary context.

ilovepi committed:
fix formatting

ilovepi committed:
Introduce helper function and remove debugging code

ilovepi committed:
Handle missing MD case

ilovepi committed:
Put comment back in original place

ilovepi committed:
Move fields to end of struct definition

ilovepi committed:
Avoid concatenation by using StringRef + Twine in calls

ilovepi committed:
git clang-format

ilovepi committed:
Avoid string operations in HTMLGenerator

ilovepi commented:
### Merge activity

* **Apr 18, 9:06 PM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/135457).
